### PR TITLE
Kmap

### LINKED ISSUE #3889: DeprecationWarning from numpy in sparse/
This line gets it: https://github.com/Theano/Theano/blob/master/theano/sparse/basic.py#L898

Simple pdb tracing reveals that in one case, grad is of shape (64,) and gout_data is also of shape (64,), but self.kmap is a list of length 36.  So there is something fishy there.

However the tests currently pass so I'm not sure what exactly is wrong.

### PR DESCRIPTION
**Pull request by shabanian**
Fixed #3889

### CODE REVIEW

**Comment by shabanian** [comment]
@adri-romsor

**Comment by nouiz** [comment]
travis is failing with at least those errors:

/home/travis/build/Theano/Theano/theano/sparse/basic.py:530:17: E711 comparison to None should be 'if cond is None:'

/home/travis/build/Theano/Theano/theano/sparse/basic.py:843:17: E711 comparison to None should be 'if cond is None:'

/home/travis/build/Theano/Theano/theano/sparse/basic.py:893:1: W293 blank line contains whitespace

======================================================================

ERROR: test_CSMGrad (theano.sparse.sandbox.test_sp.TestSP)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/Theano/Theano/theano/sparse/sandbox/test_sp.py", line 365, in test_CSMGrad

    utt.verify_grad(d, [kvals])

  File "/home/travis/build/Theano/Theano/theano/tests/unittest_tools.py", line 83, in verify_grad

    T.verify_grad(op, pt, n_tests, rng, *args, **kwargs)

  File "/home/travis/build/Theano/Theano/theano/gradient.py", line 1651, in verify_grad

    o_output = fun(*tensor_pt)

  File "/home/travis/build/Theano/Theano/theano/sparse/sandbox/test_sp.py", line 361, in d

    theano.sparse.CSM(sptype, kmap)(

  File "/home/travis/build/Theano/Theano/theano/sparse/basic.py", line 665, in __init__

    raise Exception("Do not use kmap, it is removed")

Exception: Do not use kmap, it is removed

======================================================================

ERROR: test_multilayer_sparse (theano.sparse.sandbox.test_sp.TestSP)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/Theano/Theano/theano/sparse/sandbox/test_sp.py", line 247, in test_multilayer_sparse

    nkerns[0], input, imshp, ss, mode=conv_mode)

  File "/home/travis/build/Theano/Theano/theano/sparse/sandbox/sp.py", line 330, in applySparseFilter

    sparsew = theano.sparse.CSM(sptype, kmap)(kerns, indices,

  File "/home/travis/build/Theano/Theano/theano/sparse/basic.py", line 665, in __init__

    raise Exception("Do not use kmap, it is removed")

Exception: Do not use kmap, it is removed

======================================================================

ERROR: test_sparse (theano.sparse.sandbox.test_sp.TestSP)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/Theano/Theano/theano/sparse/sandbox/test_sp.py", line 162, in test_sparse

    nkern, input, imshp, ss, bias=bias, mode=conv_mode)

  File "/home/travis/build/Theano/Theano/theano/sparse/sandbox/sp.py", line 330, in applySparseFilter

    sparsew = theano.sparse.CSM(sptype, kmap)(kerns, indices,

  File "/home/travis/build/Theano/Theano/theano/sparse/basic.py", line 665, in __init__

    raise Exception("Do not use kmap, it is removed")

Exception: Do not use kmap, it is removed

----------------------------------------------------------------------

**Comment by abergeron** [comment]
Apparently `kmap` is used for unshared sparse convolutions (only), which is in the sparse sandbox.  This is probably not something that we want to support going forward so we can remove it.

The things to remove would be `sparse_eval` at https://github.com/Theano/Theano/blob/master/theano/sparse/sandbox/sp.py#L47 and everything that depends on it.

This would also make the parameter `ws` (https://github.com/Theano/Theano/blob/master/theano/sparse/sandbox/sp.py#L76) obsolete so we should just ensure that it is always `True`.

**Comment by nouiz** [comment]
travis is failing:

```
======================================================================

ERROR: test_CSMGrad (theano.sparse.sandbox.test_sp.TestSP)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/Theano/Theano/theano/sparse/sandbox/test_sp.py", line 356, in test_CSMGrad

    sp.convolution_indices.sparse_eval(imshp, kshp, nkern, ss, conv_mode)

AttributeError: 'ConvolutionIndices' object has no attribute 'sparse_eval'

======================================================================

ERROR: test_multilayer_sparse (theano.sparse.sandbox.test_sp.TestSP)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/Theano/Theano/theano/sparse/sandbox/test_sp.py", line 247, in test_multilayer_sparse

    nkerns[0], input, imshp, ss, mode=conv_mode)

  File "/home/travis/build/Theano/Theano/theano/sparse/sandbox/sp.py", line 324, in applySparseFilter

    convolution_indices.sparse_eval(imgshp, kshp, nkern, step, mode)

AttributeError: 'ConvolutionIndices' object has no attribute 'sparse_eval'

======================================================================

ERROR: test_sparse (theano.sparse.sandbox.test_sp.TestSP)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/Theano/Theano/theano/sparse/sandbox/test_sp.py", line 162, in test_sparse

    nkern, input, imshp, ss, bias=bias, mode=conv_mode)

  File "/home/travis/build/Theano/Theano/theano/sparse/sandbox/sp.py", line 324, in applySparseFilter

    convolution_indices.sparse_eval(imgshp, kshp, nkern, step, mode)

AttributeError: 'ConvolutionIndices' object has no attribute 'sparse_eval'

----------------------------------------------------------------------
```

**Comment by adri-romsor** [comment]
@nouiz @abergeron @lamblin

File: theano/sparse/basic.py [hunk 1]
```diff
@@ -655,31 +626,22 @@ def __init__(self, format, kmap=None):
...
+        self.view_map = {0: [0]}
 
-        if not isinstance(self.kmap, numpy.ndarray):
-            # should view the other inputs too, but viewing multiple
-            # inputs is not currently supported by the destroyhandler
-            self.view_map = {0: [0]}
-
-        self._hashval = (hash(type(self)) ^ hash(self.format) ^
-                         _kmap_hash(self.kmap))
+        self._hashval = (hash(type(self)) ^ hash(self.format))
```
Comment (nouiz) [inline theano/sparse/basic.py]: It is good. Can you go one step futur and add a \__props__ = ('format',) to this op. Then you can clean it up by removing \_hashval, \__eq__, \__hash__ and \__str__.

**Comment by shabanian** [comment]
It is done! :)

File: theano/sparse/basic.py [hunk 2]
```diff
@@ -655,32 +626,14 @@ def __init__(self, format, kmap=None):
...
             raise ValueError("format must be one of: 'csr', 'csc'", format)
         self.format = format
-
-        # for efficiency, if remap does nothing, then do not apply it
-        if kmap is not None and all(kmap == numpy.arange(numpy.size(kmap))):
-            kmap = None
-
+        if kmap is not None:
+            raise Exception("Do not use kmap, it is removed")
         self.kmap = kmap
```
Comment (abergeron) [inline theano/sparse/basic.py]: You should remove this too.

File: theano/sparse/basic.py [hunk 3]
```diff
@@ -655,32 +626,14 @@ def __init__(self, format, kmap=None):
...
-                _kmap_eq(self.kmap, other.kmap))
-
-    def __hash__(self):
-        return self._hashval
-
-    def __str__(self):
-        if self.kmap is not None:
-            return "%s{%s}" % (self.__class__.__name__, str(self.kmap))
-        return self.__class__.__name__
+        self._hashval = (hash(type(self)) ^ hash(self.format))
```
Comment (abergeron) [inline theano/sparse/basic.py]: And this

**Comment by abergeron** [comment]
I think me and fred have the same comments, but ok otherwise.

**Comment by shabanian** [comment]
Done:)

**Comment by nouiz** [comment]
Travis is failing with this error:

```

File "/home/travis/build/Theano/Theano/theano/sparse/basic.py", line 705,
in grad

g_data = csm_grad(self.kmap)(x_data, x_indices, x_indptr, x_shape,

AttributeError: 'CSM' object has no attribute 'kmap'

```

So it seam that CSMGrad also need this clean up. To finish this PR that
isn't related to kamp, you can just don't pass the kmap parameter, it will
default to None. Which should allow to merge this PR. If you have the time,
you could also remove it from CSMGrad as you did for CSM.

File: theano/tensor/tests/test_opt.py [hunk 4]
```diff
@@ -6125,6 +6126,38 @@ def test3(self):
...
+
+    f = function([x], y)
+    g = function([x], z)
+    h = function([x], t)
+    r = function([u], s)
+    x_val = numpy.random.rand(4, 3).astype(config.floatX)
+    f_val = f(x_val)
+    f_test = function([x], T.expm1(x))
+
+    assert numpy.all(f_val == f_test(x_val))
```
Comment (nouiz) [inline theano/tensor/tests/test_opt.py]: use "assert numpy.allclose(f_val, f_test(x_val))". this should fix travis.

**Comment by nouiz** [comment]
instead of making a new commit that remove the commit that should not be there, can you rebase and remove the commit that should not be there? I can help if needed.

**Comment by nouiz** [comment]
thanks

**Pull request merged**